diff -rupN linux-3.10.0-1160.49.1.el7.x86_64/kernel/aes-mac1.h aes/kernel/aes-mac1.h
--- linux-3.10.0-1160.49.1.el7.x86_64/kernel/aes-mac1.h	1969-12-31 19:00:00.000000000 -0500
+++ aes/kernel/aes-mac1.h	2022-07-11 21:48:11.259787055 -0400
@@ -0,0 +1,27 @@
+/*
+*/
+
+
+
+
+#ifndef AES_MAC1_H
+#define AES_MAC1_H
+
+
+#if defined(__cplusplus)
+extern "C" {
+#endif
+
+  typedef __m128i block;
+  typedef struct {block rd_key[11]; } AES_KEY;
+ 
+
+  /* Streaming API */
+  void AES_128_Key_Expansion(const unsigned char *userkey, void *key);  
+
+#if defined(__cplusplus)
+}
+#endif
+
+#endif
+
diff -rupN linux-3.10.0-1160.49.1.el7.x86_64/kernel/audit.c aes/kernel/audit.c
--- linux-3.10.0-1160.49.1.el7.x86_64/kernel/audit.c	2021-11-09 08:29:30.000000000 -0500
+++ aes/kernel/audit.c	2022-07-11 21:47:58.757656523 -0400
@@ -66,6 +66,19 @@
 
 #include "audit.h"
 
+/* QuickLog2 integrity */
+#include <linux/random.h>
+#include <linux/types.h>
+#include <asm/i387.h>
+
+#define  _MM_MALLOC_H_INCLUDED
+#include <x86intrin.h>
+#undef _MM_MALLOC_H_INCLUDED
+
+#include "aes-mac1.h"
+//---------------------------------------------------------------------------------------
+
+
 /* No auditing will take place until audit_initialized == AUDIT_INITIALIZED.
  * (Initialization happens after skb_init is called.) */
 #define AUDIT_DISABLED		-1
@@ -179,6 +192,359 @@ struct audit_reply {
 	struct sk_buff *skb;
 };
 
+
+//QuickLog2 integrity
+//---------------------------------------------------------------------------------------
+
+//static DEFINE_SPINLOCK(lock_set_logging);
+#if 0
+typedef __m128i block;
+typedef struct {block rd_key[11]; } AES_KEY;
+#endif 
+
+const static unsigned char aeskey[16] = {0};
+static AES_KEY const_aeskey;
+static block current_key, current_state;
+static uint64_t update_one[2];
+static block *one = ((block *)(update_one));
+block quick_tag;
+
+
+/* Some helper functions */
+#define rnds 10 //AES rounds
+#define xor_block(x,y)        _mm_xor_si128(x,y)
+#define zero_block()          _mm_setzero_si128()
+/*Load 2-byte counter following 14-byte log data*/
+#define gen_logging_blk(log,ctr) _mm_insert_epi16(_mm_loadu_si128(log), ctr, 0)
+
+// Sources:
+//
+#define EXPAND_ASSIST(v1,v2,v3,v4,shuff_const,aes_const)                    \
+  do{                                                                       \
+    v2 = _mm_aeskeygenassist_si128(v4,aes_const);                           \
+    v3 = _mm_castps_si128(_mm_shuffle_ps(_mm_castsi128_ps(v3),              \
+                                         _mm_castsi128_ps(v1), 16));        \
+    v1 = _mm_xor_si128(v1,v3);                                              \
+    v3 = _mm_castps_si128(_mm_shuffle_ps(_mm_castsi128_ps(v3),              \
+                                         _mm_castsi128_ps(v1), 140));       \
+    v1 = _mm_xor_si128(v1,v3);                                              \
+    v2 = _mm_shuffle_epi32(v2,shuff_const);                                 \
+    v1 = _mm_xor_si128(v1,v2);                                              \
+  } while(0)
+
+
+void AES_128_Key_Expansion(const unsigned char *userkey, void *key)
+{
+    __m128i x0,x1,x2;
+    __m128i *kp = (__m128i *)key;
+    kp[0] = x0 = _mm_loadu_si128((__m128i*)userkey);
+    x2 = zero_block();
+    EXPAND_ASSIST(x0,x1,x2,x0,255,1);   kp[1]  = x0;
+    EXPAND_ASSIST(x0,x1,x2,x0,255,2);   kp[2]  = x0;
+    EXPAND_ASSIST(x0,x1,x2,x0,255,4);   kp[3]  = x0;
+    EXPAND_ASSIST(x0,x1,x2,x0,255,8);   kp[4]  = x0;
+    EXPAND_ASSIST(x0,x1,x2,x0,255,16);  kp[5]  = x0;
+    EXPAND_ASSIST(x0,x1,x2,x0,255,32);  kp[6]  = x0;
+    EXPAND_ASSIST(x0,x1,x2,x0,255,64);  kp[7]  = x0;
+    EXPAND_ASSIST(x0,x1,x2,x0,255,128); kp[8]  = x0;
+    EXPAND_ASSIST(x0,x1,x2,x0,255,27);  kp[9]  = x0;
+    EXPAND_ASSIST(x0,x1,x2,x0,255,54);  kp[10] = x0;
+}
+#undef EXPAND_ASSIST
+
+
+//-------------------------------------------------------------------
+
+
+#if 1
+#define gen_7_blks(cipher_blks,log_msg,counter)                            \
+    do{                                                                    \
+		cipher_blks[1]  = gen_logging_blk((block*)(log_msg+12),counter+2); \
+		cipher_blks[2]  = gen_logging_blk((block*)(log_msg+26),counter+3); \
+		cipher_blks[3]  = gen_logging_blk((block*)(log_msg+40),counter+4); \
+		cipher_blks[4]  = gen_logging_blk((block*)(log_msg+54),counter+5); \
+		cipher_blks[5]  = gen_logging_blk((block*)(log_msg+68),counter+6); \
+		cipher_blks[6]  = gen_logging_blk((block*)(log_msg+82),counter+7); \
+		cipher_blks[7]  = gen_logging_blk((block*)(log_msg+96),counter+8); \
+	} while(0)
+
+
+//AES pre-round 8 blocks
+#define prernd_8(cipher_blks, key)                   \
+	do{                                                  \
+		cipher_blks[0] = xor_block(cipher_blks[0], key); \
+		cipher_blks[1] = xor_block(cipher_blks[1], key); \
+		cipher_blks[2] = xor_block(cipher_blks[2], key); \
+		cipher_blks[3] = xor_block(cipher_blks[3], key); \
+		cipher_blks[4] = xor_block(cipher_blks[4], key); \
+		cipher_blks[5] = xor_block(cipher_blks[5], key); \
+		cipher_blks[6] = xor_block(cipher_blks[6], key); \
+		cipher_blks[7] = xor_block(cipher_blks[7], key); \
+	} while(0)
+	
+//AES Pre-round 4 blocks
+#define prernd_4(cipher_blks,sign_keys) 			       \
+  do{                                                      \
+    cipher_blks[0] = xor_block(cipher_blks[0], sign_keys); \
+    cipher_blks[1] = xor_block(cipher_blks[1], sign_keys); \
+    cipher_blks[2] = xor_block(cipher_blks[2], sign_keys); \
+    cipher_blks[3] = xor_block(cipher_blks[3], sign_keys); \
+  } while(0)
+
+
+//XOR 8 cipher blocks
+#define tag_8_xor(tag_blks,cipher_blks)                                                                       \
+  do{                                                                                                              \
+	tag_blks[0] =xor_block(xor_block(cipher_blks[0], cipher_blks[1]), xor_block(cipher_blks[2], cipher_blks[3]));  \
+	tag_blks[1] =xor_block(xor_block(cipher_blks[4], cipher_blks[5]), xor_block(cipher_blks[6], cipher_blks[7]));  \
+	tag_blks[2] =xor_block(tag_blks[2], xor_block(tag_blks[0], tag_blks[1]));                                      \
+  } while(0)
+
+
+
+
+static inline void AES_single(block *cipher_blks, block *sched)
+{
+    uint8_t j;
+	for(j=0;j<10;j++){
+		* cipher_blks = _mm_aesenc_si128(*cipher_blks, sched[j]); 
+	}	
+	*cipher_blks =_mm_aesenclast_si128(*cipher_blks, sched[10]);		
+}
+
+
+
+
+
+static inline void AES_ECB_2(block *cipher_blks, block *sched)
+{
+    uint8_t j;
+	for(j=0;j<10;j++){
+		cipher_blks[0] = _mm_aesenc_si128(cipher_blks[0], sched[j]); 
+		cipher_blks[1] = _mm_aesenc_si128(cipher_blks[1], sched[j]); 
+	}	
+	cipher_blks[0] =_mm_aesenclast_si128(cipher_blks[0], sched[10]);
+	cipher_blks[1] =_mm_aesenclast_si128(cipher_blks[1], sched[10]);		
+}
+
+
+
+static inline void AES_ECB_4(block *cipher_blks, block *sched)
+{
+    uint8_t j;
+	for(j=0;j<10;j++){
+		cipher_blks[0] = _mm_aesenc_si128(cipher_blks[0], sched[j]); 
+		cipher_blks[1] = _mm_aesenc_si128(cipher_blks[1], sched[j]); 
+		cipher_blks[2] = _mm_aesenc_si128(cipher_blks[2], sched[j]); 
+		cipher_blks[3] = _mm_aesenc_si128(cipher_blks[3], sched[j]); 
+	}	
+	cipher_blks[0] =_mm_aesenclast_si128(cipher_blks[0], sched[10]);
+	cipher_blks[1] =_mm_aesenclast_si128(cipher_blks[1], sched[10]);		
+	cipher_blks[2] =_mm_aesenclast_si128(cipher_blks[2], sched[10]);
+	cipher_blks[3] =_mm_aesenclast_si128(cipher_blks[3], sched[10]);
+}
+
+
+
+
+static inline void AES_ECB_8(block *cipher_blks, block *sched)
+{
+    uint8_t j;
+	for(j=0;j<10;j++){
+		cipher_blks[0] = _mm_aesenc_si128(cipher_blks[0], sched[j]); 
+		cipher_blks[1] = _mm_aesenc_si128(cipher_blks[1], sched[j]); 
+		cipher_blks[2] = _mm_aesenc_si128(cipher_blks[2], sched[j]); 
+		cipher_blks[3] = _mm_aesenc_si128(cipher_blks[3], sched[j]); 
+		cipher_blks[4] = _mm_aesenc_si128(cipher_blks[4], sched[j]); 
+		cipher_blks[5] = _mm_aesenc_si128(cipher_blks[5], sched[j]); 
+		cipher_blks[6] = _mm_aesenc_si128(cipher_blks[6], sched[j]); 
+		cipher_blks[7] = _mm_aesenc_si128(cipher_blks[7], sched[j]);
+	}	
+	cipher_blks[0] =_mm_aesenclast_si128(cipher_blks[0], sched[10]);
+	cipher_blks[1] =_mm_aesenclast_si128(cipher_blks[1], sched[10]);		
+	cipher_blks[2] =_mm_aesenclast_si128(cipher_blks[2], sched[10]);
+	cipher_blks[3] =_mm_aesenclast_si128(cipher_blks[3], sched[10]);
+	cipher_blks[4] =_mm_aesenclast_si128(cipher_blks[4], sched[10]);
+	cipher_blks[5] =_mm_aesenclast_si128(cipher_blks[5], sched[10]);
+	cipher_blks[6] =_mm_aesenclast_si128(cipher_blks[6], sched[10]);
+	cipher_blks[7] =_mm_aesenclast_si128(cipher_blks[7], sched[10]);
+}
+
+
+static inline void first_blks(block *cipher_blks, uint16_t counter, char *log_msg, block *sched, block sign_keys)
+{
+		//uint8_t i;
+		cipher_blks[0]  = _mm_srli_si128(_mm_loadu_si128((block*)log_msg), 2); 
+		cipher_blks[0]  = _mm_insert_epi16(cipher_blks[0], counter+1, 0);
+		gen_7_blks(cipher_blks,log_msg,counter);
+		/*AES Preround*/
+		prernd_8(cipher_blks,sign_keys);
+		/*AES Round 1-9*/	
+		//for(i=1; i<rnds; ++i) aes_enc_8(cipher_blks,sched, i);			
+		/*Last Round(10)*/
+		//aes_enclast_8(cipher_blks, sched[rnds]);
+		AES_ECB_8(cipher_blks, sched);		
+}
+
+ static inline void cmpt_8_blks(block *cipher_blks, uint16_t counter, char *log_msg, block *sched, block sign_keys)
+ {
+	//uint8_t i;
+	cipher_blks[0]  = gen_logging_blk((block*)log_msg, counter+1); 
+	gen_7_blks(cipher_blks,log_msg,counter);
+	/*AES Preround */
+	prernd_8(cipher_blks, sign_keys);
+	/*AES Round 1-9*/
+	//for (i=0; i<rnds; ++i) aes_enc_8(cipher_blks, sched, i);		
+	/*AES Last Round(10)*/
+	AES_ECB_8(cipher_blks, sched);	
+}
+
+
+static inline void cmpt_4_blks(block *cipher_blks, uint16_t counter, const char *log_msg, block *sched, block sign_keys)
+{
+    	//uint8_t i;
+		if(counter){		
+			cipher_blks[0]  = gen_logging_blk((block*)(log_msg),counter+1); 
+		}else{//contains the first block
+			cipher_blks[0]  = _mm_srli_si128(_mm_loadu_si128((block*)log_msg), 2);
+			cipher_blks[0]  = _mm_insert_epi16(cipher_blks[0], counter+1, 0);
+		}
+		cipher_blks[1]  = gen_logging_blk((block*)(log_msg+12),counter+2); 
+		cipher_blks[2]  = gen_logging_blk((block*)(log_msg+26),counter+3); 
+		cipher_blks[3]  = gen_logging_blk((block*)(log_msg+40),counter+4); 
+		/*AES Preround*/
+		prernd_4(cipher_blks,sign_keys);
+		AES_ECB_4(cipher_blks, sched);
+}
+
+
+static inline void cmpt_2_blks(block *cipher_blks, uint16_t counter, const char *log_msg, block *sched, block sign_keys)
+{
+    	
+		 if(counter){		
+			cipher_blks[0]  = gen_logging_blk((block*)(log_msg),counter+1); //Not the first block
+		}else{//contains the first block
+			cipher_blks[0]  = _mm_srli_si128(_mm_loadu_si128((block*)log_msg), 2);
+			cipher_blks[0]  = _mm_insert_epi16(cipher_blks[0], counter+1, 0);
+		}
+		cipher_blks[1]  = gen_logging_blk((block*)(log_msg+12),counter+2); 
+		/*AES Preround */
+		cipher_blks[0] =_mm_xor_si128(cipher_blks[0], sign_keys);
+		cipher_blks[1] =_mm_xor_si128(cipher_blks[1], sign_keys);
+		AES_ECB_2(cipher_blks, sched);
+}
+
+
+static inline void cmpt_a_blk(block* cipher_blk, uint16_t counter, const char *log_msg, block *sched, block sign_keys)
+{
+    	
+		if(counter){
+				*cipher_blk = _mm_loadu_si128((block*)log_msg);//Not the first block
+		}else{//it is the first block
+				*cipher_blk = _mm_srli_si128(_mm_loadu_si128((block*)log_msg), 2);//the first block
+		}
+		*cipher_blk = _mm_insert_epi16(*cipher_blk, counter+1, 0);
+		
+		/*AES Preround */
+		*cipher_blk =_mm_xor_si128(*cipher_blk, sign_keys);
+		AES_single(cipher_blk, sched);
+}
+
+
+/**  
+* MAC, signing a log message and updating the signing-key & state
+* Input @log_msg: a log data, 
+* Computing block format: a block(16 bytes) contains "<i>||M_i",  
+*                         2 bytes counter(<i>) and 14 bytes log data(M_i)
+**/
+static void mac_core( char *log_msg)
+{
+   size_t msg_len = strlen(log_msg);
+	uint16_t i, remaining , counter = 0;  /*now the max log size is 8192 KB*/
+	block * sched, *aes_blks, *last;
+	block mask, update_mask, cipher_blks[9], tag_blks[3], next[2];
+	unsigned char *my_pad, *chr_cnt; /*for padding the last block*/
+	remaining = (uint16_t)msg_len;
+	sched = ((block *)(const_aeskey.rd_key)); //point to AES round keys
+	aes_blks = cipher_blks;
+	my_pad = (unsigned char*)kmalloc(32 * sizeof(unsigned char), GFP_KERNEL);
+	last = ((block *)(my_pad)); 
+	chr_cnt = (unsigned char *)&counter;
+	tag_blks[2] = current_key;
+
+	mask =_mm_xor_si128(sched[0], current_key);/*xor the signing key with the aes public key*/
+	update_mask =_mm_xor_si128(sched[0], current_state);/*xor the current state with the aes public key*/
+	
+	if(remaining>=112)//start 8 blocks parallel computing 
+	{
+		first_blks(aes_blks, counter, log_msg, sched, mask);
+		tag_8_xor(tag_blks,aes_blks);
+		counter +=8;
+		log_msg +=110;/*112-byte computed, apply 110-byte, leaving 2-byte overwrote by counter*/
+		remaining -= 112;
+		while(remaining >= 112){
+			cmpt_8_blks(aes_blks, counter, log_msg, sched, mask);
+			tag_8_xor(tag_blks,aes_blks);
+			counter +=8;
+			log_msg +=110;
+			remaining -= 112;
+		}
+	}//end of nblks	
+    if (remaining){
+		if(remaining >=56){//4-block, 4*14=56 bytes log data
+			cmpt_4_blks(aes_blks,counter, log_msg, sched, mask);
+			tag_blks[0] = xor_block( xor_block(aes_blks[0], aes_blks[1]), xor_block(aes_blks[2], aes_blks[3])); 
+			tag_blks[2] = xor_block(tag_blks[2], tag_blks[0]);
+			remaining -= 56;
+			counter +=4;
+			log_msg +=54;/*56-byte computed, apply 54-byte, leaving 2-byte overwrote by counter*/
+		}
+		if (remaining >= 28) {//2-block, 2*14=28 bytes log data
+			cmpt_2_blks(aes_blks,counter, log_msg, sched, mask);
+			tag_blks[2] = xor_block(xor_block(aes_blks[0], aes_blks[1]), tag_blks[2]); 
+			remaining -= 28;
+			counter +=2;
+			log_msg +=26;/*28-byte computed, apply 26-byte, leaving 2-byte overwrote by counter*/
+		}
+		if (remaining >= 14) {//1-block 14 bytes log data
+			cmpt_a_blk(&aes_blks[0],counter, log_msg, sched, mask);
+			tag_blks[2] = xor_block(tag_blks[2], aes_blks[0]);
+			remaining -= 14;
+			counter +=1;
+			log_msg +=12;/*14-byte computed, apply 12-byte, leaving 2-byte overwrote by counter*/
+		}
+		if (remaining){
+			if (counter)  log_msg +=2;
+			counter += (14-remaining);
+			*last = zero_block();
+			my_pad[0] = chr_cnt[0];
+			my_pad[1] = chr_cnt[1];
+			for(i=0;i<remaining;i++) my_pad[i+2]=log_msg[i];
+			aes_blks[0] = xor_block(*last, mask);
+			AES_single(&aes_blks[0], sched);
+			tag_blks[2] = xor_block(tag_blks[2], aes_blks[0]);
+		}
+    }//end of remaining
+	next[0] = update_mask;
+	next[1] = xor_block(*one, update_mask);
+	AES_ECB_2(next, sched);
+	current_key = xor_block(next[0], current_state);
+	current_state = xor_block(next[1], current_state);
+	quick_tag = xor_block(quick_tag, tag_blks[2]);
+}
+
+
+#undef gen_7_blks
+#undef prernd_8
+#undef prernd_4
+#undef tag_8_xor
+#endif
+
+//-----------------------------------------------------------------
+//endregion
+
+
+
 static void audit_set_portid(struct audit_buffer *ab, __u32 portid)
 {
 	if (ab) {
@@ -1076,7 +1442,11 @@ static void audit_receive(struct sk_buff
 /* Initialize audit support at boot time. */
 static int __init audit_init(void)
 {
-	int i;
+	int i; 
+	/* QuickLog2 integrity */
+	block s_0, mask;
+	block init_pair[2];
+	block * sched;
 	struct netlink_kernel_cfg cfg = {
 		.input	= audit_receive,
 	};
@@ -1100,7 +1470,25 @@ static int __init audit_init(void)
 
 	for (i = 0; i < AUDIT_INODE_BUCKETS; i++)
 		INIT_LIST_HEAD(&audit_inode_hash[i]);
+	
+	//---------------------------------------------------------------------------------------
+	/* QuickLog2 integrity */
+	get_random_bytes(&s_0, sizeof(block));/*initial State */
+	kernel_fpu_begin();	
+	AES_128_Key_Expansion(aeskey,&const_aeskey); //inital aes round keys
+	init_pair[0] = zero_block();/*0 for updatting state*/
+	init_pair[1] = _mm_setr_epi32(0x0001, 0x0000, 0x0000, 0x0000);/*1 for updatting key*/
+	sched = ((block *)(const_aeskey.rd_key)); 
+	mask =xor_block(s_0, sched[0]);
+	init_pair[0] = xor_block(init_pair[0], mask);
+	init_pair[1] = xor_block(init_pair[1], mask);
+	AES_ECB_2(init_pair, sched);
+	current_state = xor_block(init_pair[0], s_0);
+	current_key = xor_block(init_pair[1], s_0);
+	quick_tag = zero_block();
+	kernel_fpu_end();
 
+	//---------------------------------------------------------------------------------------
 	return 0;
 }
 __initcall(audit_init);
@@ -1895,20 +2283,38 @@ out:
  */
 void audit_log_end(struct audit_buffer *ab)
 {
+	struct nlmsghdr *nlh;
+	char *log_msg; // log message
+       unsigned long flags;
+	
 	if (!ab)
 		return;
 	if (!audit_rate_check()) {
 		audit_log_lost("rate limit exceeded");
 	} else {
-		struct nlmsghdr *nlh = nlmsg_hdr(ab->skb);
-		nlh->nlmsg_len = ab->skb->len - NLMSG_HDRLEN;
+		nlh = nlmsg_hdr(ab->skb);
 
-		if (audit_pid) {
-			skb_queue_tail(&audit_skb_queue, ab->skb);
-			wake_up_interruptible(&kauditd_wait);
-		} else {
-			audit_printk_skb(ab->skb);
-		}
+//------------QuickLog2 start -------------
+	log_msg = nlmsg_data(nlh);
+
+		
+/* Perform the cryptographic operations synchronously */
+	spin_lock_irqsave(&(&audit_skb_queue)->lock, flags);
+	kernel_fpu_begin();
+	mac_core(log_msg);
+	kernel_fpu_end();
+		
+	nlh->nlmsg_len = ab->skb->len - NLMSG_HDRLEN;
+
+	if (audit_pid) {
+		__skb_queue_tail(&audit_skb_queue, ab->skb);
+		spin_unlock_irqrestore(&(&audit_skb_queue)->lock, flags);
+		wake_up_interruptible(&kauditd_wait);
+	} else {
+		spin_unlock_irqrestore(&(&audit_skb_queue)->lock, flags);
+		audit_printk_skb(ab->skb);
+	}
+//----------------------------------
 		ab->skb = NULL;
 	}
 	audit_buffer_free(ab);
diff -rupN linux-3.10.0-1160.49.1.el7.x86_64/kernel/Makefile aes/kernel/Makefile
--- linux-3.10.0-1160.49.1.el7.x86_64/kernel/Makefile	2021-11-09 08:29:30.000000000 -0500
+++ aes/kernel/Makefile	2022-07-11 21:48:37.241058321 -0400
@@ -2,6 +2,8 @@
 # Makefile for the linux kernel.
 #
 
+ccflags-y := -Wno-declaration-after-statement -mmmx -maes -mpreferred-stack-boundary=4
+
 obj-y     = fork.o exec_domain.o panic.o printk.o \
 	    cpu.o exit.o itimer.o time.o softirq.o resource.o \
 	    sysctl.o sysctl_binary.o capability.o ptrace.o timer.o user.o \
